File: native/cocos/renderer/renderer/ForwardRenderer.cpp
/****************************************************************************
 Copyright (c) 2018 Xiamen Yaji Software Co., Ltd.

 http://www.cocos2d-x.org
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#include "ForwardRenderer.h"

#include "gfx/DeviceGraphics.h"
#include "gfx/Texture2D.h"
#include "gfx/VertexBuffer.h"
#include "gfx/IndexBuffer.h"
#include "gfx/FrameBuffer.h"
#include "ProgramLib.h"
#include "View.h"
#include "Scene.h"
#include "Effect.h"
#include "InputAssembler.h"
#include "Pass.h"
#include "Camera.h"
#include "Light.h"
#include <algorithm>

RENDERER_BEGIN

ForwardRenderer::ForwardRenderer()
{
    _arrayPool = new RecyclePool<float>([]()mutable->float*{return new float[16];}, 8);

}

ForwardRenderer::~ForwardRenderer()
{
    _directionalLights.clear();
    _pointLights.clear();
    _spotLights.clear();
    _shadowLights.clear();
    _ambientLights.clear();
    _defines.clear();
    
    delete _arrayPool;
    _arrayPool = nullptr;
}

bool ForwardRenderer::init(DeviceGraphics* device, std::vector<ProgramLib::Template>& programTemplates, Texture2D* defaultTexture, int width, int height)
{
    BaseRenderer::init(device, programTemplates, defaultTexture);
    _width = width;
    _height = height;
    registerStage("opaque", std::bind(&ForwardRenderer::opaqueStage, this, std::placeholders::_1, std::placeholders::_2));
    registerStage("shadowcast", std::bind(&ForwardRenderer::shadowStage, this, std::placeholders::_1, std::placeholders::_2));
    registerStage("transparent", std::bind(&ForwardRenderer::transparentStage, this, std::placeholders::_1, std::placeholders::_2));
    return true;
}

void ForwardRenderer::resetData()
{
    _arrayPool->reset();
    reset();
}

void ForwardRenderer::render(Scene* scene)
{
    resetData();
    updateLights(scene);
    scene->sortCameras();
    auto& cameras = scene->getCameras();
    for (auto& camera : cameras)
    {
        View* view = requestView();
        camera->extractView(*view, _width, _height);
    }

    for (size_t i = 0, len = _views->getLength(); i < len; ++i) {
        const View* view = _views->getData(i);
        BaseRenderer::render(*view, scene);
    }
    
    scene->removeModels();
}

void ForwardRenderer::renderCamera(Camera* camera, Scene* scene)
{
    reset();
    int width = _width;
    int height = _height;
    FrameBuffer* fb = camera->getFrameBuffer();
    if (nullptr != fb) {
        width = fb->getWidth();
        height = fb->getHeight();
    }
    View* view = requestView();
    camera->extractView(*view, width, height);
    BaseRenderer::render(*view, scene);
    
    scene->removeModels();
}

void ForwardRenderer::updateLights(Scene* scene)
{
    _directionalLights.clear();
    _pointLights.clear();
    _spotLights.clear();
    _shadowLights.clear();
    _ambientLights.clear();
    
    const Vector<Light*> lights = scene->getLights();
    for (const auto& light : lights)
    {
        light->update(_device);
        if (light->getShadowType() != Light::ShadowType::NONE)
        {
            _shadowLights.pushBack(light);
            View* view = requestView();
            std::vector<std::string> stages;
            stages.push_back("shadowcast");
            light->extractView(*view, stages);
        }
        
        switch(light->getType())
        {
            case Light::LightType::DIRECTIONAL:
            {
                _directionalLights.pushBack(light);
                break;
            }
            case Light::LightType::POINT:
            {
                _pointLights.pushBack(light);
                break;
            }
            case Light::LightType::SPOT:
            {
                _spotLights.pushBack(light);
                break;
            }
            case Light::LightType::AMBIENT:
            {
                _ambientLights.pushBack(light);
                break;
            }
        }
    }
    
    if (lights.size() > 0)
    {
        updateDefines();
    }
    
    _numLights = lights.size();
}

void ForwardRenderer::updateDefines()
{
    _defines.emplace(std::make_pair("_NUM_DIR_LIGHTS", Value(std::min(4, (int)_directionalLights.size()))));
    _defines.emplace(std::make_pair("_NUM_POINT_LIGHTS", Value(std::min(4, (int)_pointLights.size()))));
    _defines.emplace(std::make_pair("_NUM_SPOT_LIGHTS", Value(std::min(4, (int)_spotLights.size()))));
    _defines.emplace(std::make_pair("_NUM_AMBIENT_LIGHTS", Value(std::min(4, (int)_ambientLights.size()))));
    _defines.emplace(std::make_pair("CC_NUM_SHADOW_LIGHTS", Value(std::min(4, (int)_shadowLights.size()))));
}

void ForwardRenderer::submitLightsUniforms()
{
    if (_directionalLights.size() > 0)
    {
        size_t count = _directionalLights.size();
        float* directions = _arrayPool->add();
        float* colors = _arrayPool->add();
        Vec3 lightVec3;
        Vec3 colorVec3;
        for (int i = 0; i < count; ++i)
        {
            int index = i * 4;
            auto* light = _directionalLights.at(i);
            lightVec3.set(light->getDirectionUniform());
            *(directions + index) = lightVec3.x;
            *(directions + index + 1) = lightVec3.y;
            *(directions + index + 2) = lightVec3.z;
            *(directions + index + 3) = 0;
            
            colorVec3.set(light->getColorUniform());
            *(colors + index) = colorVec3.x;
            *(colors + index + 1) = colorVec3.y;
            *(colors + index + 2) = colorVec3.z;
            *(colors + index + 3) = 0;
        }
        _device->setUniformfv("cc_dirLightDirection", count * 4, directions);
        _device->setUniformfv("cc_dirLightColor", count * 4, colors);
    }
    
    if (_pointLights.size() > 0)
    {
        size_t count = _pointLights.size();
        float* positionAndRanges = _arrayPool->add();
        float* colors = _arrayPool->add();
        Vec3 posVec3;
        Vec3 colorVec3;
        for (int i = 0; i < count; ++i)
        {
            int index = i * 4;
            auto* light = _pointLights.at(i);
            posVec3.set(light->getPositionUniform());
            *(positionAndRanges + index) = posVec3.x;
            *(positionAndRanges + index + 1) = posVec3.y;
            *(positionAndRanges + index + 2) = posVec3.z;
            *(positionAndRanges + index + 3) = light->getRange();

            colorVec3.set(light->getColorUniform());
            *(colors + index) = colorVec3.x;
            *(colors + index + 1) = colorVec3.y;
            *(colors + index + 2) = colorVec3.z;
            *(colors + index + 3) = 0;
        }
        _device->setUniformfv("cc_pointLightPositionAndRange", count * 4, positionAndRanges);
        _device->setUniformfv("cc_pointLightColor", count * 4, colors);
    }
    
    if (_spotLights.size() > 0)
    {
        size_t count = _spotLights.size();
        float* directions = _arrayPool->add();
        float* positionAndRanges = _arrayPool->add();
        float* colors = _arrayPool->add();
        Vec3 lightVec3;
        Vec3 posVec3;
        Vec3 colorVec3;
        for (int i = 0; i < count; ++i)
        {
            int index = i * 4;
            auto* light = _spotLights.at(i);
            lightVec3.set(light->getDirectionUniform());
            *(directions + index) = lightVec3.x;
            *(directions + index + 1) = lightVec3.y;
            *(directions + index + 2) = lightVec3.z;
            *(directions + index + 3) = 0;
            
            posVec3.set(light->getPositionUniform());
            *(positionAndRanges + index) = posVec3.x;
            *(positionAndRanges + index + 1) = posVec3.y;
            *(positionAndRanges + index + 2) = posVec3.z;
            *(positionAndRanges + index + 3) = light->getRange();
            
            colorVec3.set(light->getColorUniform());
            *(colors + index) = colorVec3.x;
            *(colors + index + 1) = colorVec3.y;
            *(colors + index + 2) = colorVec3.z;
            *(colors + index + 3) = 0;
        }
        _device->setUniformfv("cc_spotLightDirection", count * 4, directions);
        _device->setUniformfv("cc_spotLightPositionAndRange", count * 4, positionAndRanges);
        _device->setUniformfv("cc_spotLightColor", count * 4, colors);
    }
    
    if (_ambientLights.size() > 0) {
        size_t count = _ambientLights.size();
        float* colors = _arrayPool->add();
        Vec3 colorVec3;
        for (int i = 0; i < count; ++i)
        {
            int index = i * 4;
            auto* light = _ambientLights.at(i);
            colorVec3.set(light->getColorUniform());
            *(colors + index) = colorVec3.x;
            *(colors + index + 1) = colorVec3.y;
            *(colors + index + 2) = colorVec3.z;
            *(colors + index + 3) = 0;
        }
        _device->setUniformfv("cc_pointLightColor", count * 4, colors);
    }
}

void ForwardRenderer::submitShadowStageUniforms(const View& view)
{
    static float* shadowInfo = new float[4];
    shadowInfo[0] = view.shadowLight->getShadowMinDepth();
    shadowInfo[1] = view.shadowLight->getShadowMaxDepth();
    shadowInfo[2] = view.shadowLight->getShadowDepthScale();
    shadowInfo[3] = view.shadowLight->getShadowDarkness();
    
    _device->setUniformMat4("cc_shadow_map_lightViewProjMatrix", view.matViewProj);
    _device->setUniformfv("cc_shadow_map_info", 4, shadowInfo);
    _device->setUniformf("cc_shadow_map_bias", view.shadowLight->getShadowBias());
}

void ForwardRenderer::submitOtherStagesUniforms()
{
    size_t count = _shadowLights.size();
    float* shadowLightInfo = _arrayPool->add();
    static float* shadowLightProjs = new float[count * 16];
    for (int i = 0; i < count; ++i)
    {
        int index = i * 4;
        
        Light* light = _shadowLights.at(i);
        const Mat4 view = light->getViewProjMatrix();
        memcpy(shadowLightProjs + index, view.m, 16);
        
        *(shadowLightInfo + index) = light->getShadowMinDepth();
        *(shadowLightInfo + index + 1) = light->getShadowMaxDepth();
        *(shadowLightInfo + index + 2) = light->getShadowDepthScale();
        *(shadowLightInfo + index + 3) = light->getShadowDarkness();
    }
    
    _device->setUniformfv("cc_shadow_lightViewProjMatrix", count * 16, shadowLightProjs);
    _device->setUniformfv("cc_shadow_info", count * 4, shadowLightInfo);
}

void ForwardRenderer::updateShaderDefines(const StageItem& item)
{
    for (auto& e : _defines)
    {
        const std::string& key = e.first;
        if(item.defines->count(key) == 0)
        {
            item.defines->emplace(key, e.second);
        }
        else
        {
            item.defines->find(key)->second = e.second;
        }
    }
}

bool ForwardRenderer::compareItems(const StageItem &a, const StageItem &b)
{
    const Technique* techA = a.technique;
    const Technique* techB = b.technique;
    
    if (techA->getLayer() != techB->getLayer()) {
        return techA->getLayer() > techB->getLayer();
    }
    
    size_t pa = techA->getPasses().size();
    size_t pb = techB->getPasses().size();
    
    if (pa != pb) {
        return pa > pb;
    }
    
    return a.sortKey > b.sortKey;
}

void ForwardRenderer::sortItems(std::vector<StageItem>& items)
{
    std::sort(items.begin(), items.end(), compareItems);
}

void ForwardRenderer::drawItems(const std::vector<StageItem>& items)
{
    size_t count = _shadowLights.size();
    if (count == 0 && _numLights == 0)
    {
        for (const auto& item : items)
        {
            draw(item);
        }
    }
    else
    {
        std::vector<Texture*> shadowMaps;
        shadowMaps.reserve(count);
        std::vector<int> slots;
        slots.reserve(count);
        for (auto item : items)
        {
            shadowMaps.clear();
            for(int i = 0; i < count; i++)
            {
                Light* light = _shadowLights.at(i);
                shadowMaps.push_back(light->getShadowMap());
                slots.push_back(allocTextureUnit());
            }
            _device->setTextureArray("cc_shadow_map", shadowMaps, slots);
            updateShaderDefines(item);
            draw(item);
        }
    }
}

void ForwardRenderer::opaqueStage(const View& view, const std::vector<StageItem>& items)
{
    // update uniforms
    _device->setUniformMat4("cc_matView", view.matView);
    _device->setUniformMat4("cc_matpProj", view.matProj);
    _device->setUniformMat4("cc_matViewProj", view.matViewProj);
    static Vec3 cameraPos3;
    static Vec4 cameraPos4;
    view.getPosition(cameraPos3);
    cameraPos4.set(cameraPos3.x, cameraPos3.y, cameraPos3.z, 0);
    _device->setUniformVec4("cc_cameraPos", cameraPos4);
    submitLightsUniforms();
    submitOtherStagesUniforms();
    drawItems(items);
}

void ForwardRenderer::shadowStage(const View& view, const std::vector<StageItem>& items)
{
    // update rendering
    submitShadowStageUniforms(view);
    
    for (auto& item : items)
    {
        if (_programLib->getValueFromDefineList("CC_SHADOW_CASTING", *item.defines) != Value::Null) {
            updateShaderDefines(item);
            draw(item);
        }
    }
}

void ForwardRenderer::transparentStage(const View& view, const std::vector<StageItem>& items)
{
    // update uniforms
    _device->setUniformMat4("cc_matView", view.matView);
    _device->setUniformMat4("cc_matpProj", view.matProj);
    _device->setUniformMat4("cc_matViewProj", view.matViewProj);
    static Vec3 cameraPos3;
    static Vec4 cameraPos4;
    view.getPosition(cameraPos3);
    cameraPos4.set(cameraPos3.x, cameraPos3.y, cameraPos3.z, 0);
    _device->setUniformVec4("cc_cameraPos", cameraPos4);
    
    static Vec3 camFwd;
    static Vec3 tmpVec3;
    view.getForward(camFwd);
    
    submitLightsUniforms();
    submitOtherStagesUniforms();
    
    NodeProxy* node;
    // calculate zdist
    for (auto& item : const_cast<std::vector<StageItem>&>(items))
    {
        // TODO: we should use mesh center instead!
        node = const_cast<NodeProxy*>(item.model->getNode());
        if (node != nullptr)
        {
            node->getWorldPosition(&tmpVec3);
        }
        else
        {
            tmpVec3.set(0, 0, 0);
        }
        
        Vec3::subtract(tmpVec3, tmpVec3, &cameraPos3);
        item.sortKey = -Vec3::dot(tmpVec3, camFwd);
    }
    
    sortItems(const_cast<std::vector<StageItem>&>(items));
    drawItems(items);
}

RENDERER_END

File: cocos2d/renderer/renderers/forward-renderer.js
// Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.

import { Vec3, Vec4, Mat4 } from '../../core/value-types';
import BaseRenderer from '../core/base-renderer';
import enums from '../enums';
import { RecyclePool } from '../memop';

let _a16_view = new Float32Array(16);
let _a16_view_inv = new Float32Array(16);
let _a16_proj = new Float32Array(16);
let _a16_viewProj = new Float32Array(16);
let _a4_camPos = new Float32Array(4);

let _a64_shadow_lightViewProj = new Float32Array(64);
let _a16_shadow_lightViewProjs = [];
let _a4_shadow_info = new Float32Array(4);

let _camPos = new Vec4(0, 0, 0, 0);
let _camFwd = new Vec3(0, 0, 0);
let _v3_tmp1 = new Vec3(0, 0, 0);

const CC_MAX_LIGHTS = 4;
const CC_MAX_SHADOW_LIGHTS = 2;

let _float16_pool = new RecyclePool(() => {
  return new Float32Array(16);
}, 8);

function sortView (a, b) {
  return (a._priority - b._priority);
}

export default class ForwardRenderer extends BaseRenderer {
  constructor(device, builtin) {
    super(device, builtin);

    this._time = new Float32Array(4);

    this._lights = [];
    this._shadowLights = [];

    this._numLights = 0;

    this._defines = {
    };

    this._registerStage('shadowcast', this._shadowStage.bind(this));
    this._registerStage('opaque', this._opaqueStage.bind(this));
    this._registerStage('transparent', this._transparentStage.bind(this));
  }

  reset () {
    _float16_pool.reset();
    super.reset();
  }

  render (scene, dt) {
    this.reset();

    if (!CC_EDITOR) {
      if (dt) {
        this._time[0] += dt;
        this._time[1] = dt;
        this._time[2] ++;
      }
      this._device.setUniform('cc_time', this._time);
    }

    this._updateLights(scene);

    const canvas = this._device._gl.canvas;
    for (let i = 0; i < scene._cameras.length; ++i) {
      let view = this._requestView();
      let width = canvas.width;
      let height = canvas.height;
      let camera = scene._cameras.data[i];
      camera.extractView(view, width, height);
    }

    // render by cameras
    this._viewPools.sort(sortView);

    for (let i = 0; i < this._viewPools.length; ++i) {
      let view = this._viewPools.data[i];
      this._render(view, scene);
    }
  }

  // direct render a single camera
  renderCamera (camera, scene) {
    this.reset();

    this._updateLights(scene);

    const canvas = this._device._gl.canvas;
    let width = canvas.width;
    let height = canvas.height;

    let view = this._requestView();
    camera.extractView(view, width, height);

    // render by cameras
    this._viewPools.sort(sortView);

    for (let i = 0; i < this._viewPools.length; ++i) {
      let view = this._viewPools.data[i];
      this._render(view, scene);
    }
  }

  _updateLights (scene) {
    this._lights.length = 0;
    this._shadowLights.length = 0;

    let lights = scene._lights;
    for (let i = 0; i < lights.length; ++i) {
      let light = lights.data[i];
      light.update(this._device);

      if (light.shadowType !== enums.SHADOW_NONE) {
        if (this._shadowLights.length < CC_MAX_SHADOW_LIGHTS) {
          this._shadowLights.splice(0, 0, light);
        }
        let view = this._requestView();
        light.extractView(view, ['shadowcast']);

        this._lights.splice(0, 0, light);
      }
      else {
        this._lights.push(light);
      }
    }

    this._updateLightDefines();
    this._numLights = lights._count;
  }

  _updateLightDefines () {
    let defines = this._defines;

    for (let i = 0; i < this._lights.length; ++i) {
      let light = this._lights[i];
      let lightKey = `CC_LIGHT_${i}_TYPE`;
      let shadowKey = `CC_SHADOW_${i}_TYPE`;
      if (defines[lightKey] !== light._type){
        defines[lightKey] = light._type;
        this._definesChanged = true;
      }
      if (defines[shadowKey] !== light._shadowType){
        defines[shadowKey] = light._shadowType;
        this._definesChanged = true;
      }
    }

    let newCount = Math.min(CC_MAX_LIGHTS, this._lights.length);
    if (defines.CC_NUM_LIGHTS !== newCount) {
      defines.CC_NUM_LIGHTS = newCount;
      this._definesChanged = true;
    }
    newCount = Math.min(CC_MAX_LIGHTS, this._shadowLights.length);
    if (defines.CC_NUM_SHADOW_LIGHTS !== newCount) {
      defines.CC_NUM_SHADOW_LIGHTS = newCount;
      this._definesChanged = true;
    }
  }

  _submitLightsUniforms () {
    let device = this._device;

    if (this._lights.length > 0) {
      let positionAndRanges = _float16_pool.add();
      let directions = _float16_pool.add();
      let colors = _float16_pool.add();
      let lightNum = Math.min(CC_MAX_LIGHTS, this._lights.length);
      for (let i = 0; i < lightNum; ++i) {
        let light = this._lights[i];
        let index = i * 4;

        colors.set(light._colorUniform, index);
        directions.set(light._directionUniform, index);
        positionAndRanges.set(light._positionUniform, index);
        positionAndRanges[index+3] = light._range;

        if (light._type === enums.LIGHT_SPOT) {
          directions[index+3] = light._spotUniform[0];
          colors[index+3] = light._spotUniform[1];
        }
        else {
          directions[index+3] = 0;
          colors[index+3] = 0;
        }
      }

      device.setUniform('cc_lightDirection', directions);
      device.setUniform('cc_lightColor', colors);
      device.setUniform('cc_lightPositionAndRange', positionAndRanges);
    }
  }

  _submitShadowStageUniforms(view) {

    let light = view._shadowLight;

    let shadowInfo = _a4_shadow_info;
    shadowInfo[0] = light.shadowMinDepth;
    shadowInfo[1] = light.shadowMaxDepth;
    shadowInfo[2] = light.shadowDepthScale;
    shadowInfo[3] = light.shadowDarkness;

    this._device.setUniform('cc_shadow_map_lightViewProjMatrix', Mat4.toArray(_a16_viewProj, view._matViewProj));
    this._device.setUniform('cc_shadow_map_info', shadowInfo);
    this._device.setUniform('cc_shadow_map_bias', light.shadowBias);

    this._defines.CC_SHADOW_TYPE = light._shadowType;
  }

  _submitOtherStagesUniforms() {
    let shadowInfo = _float16_pool.add();

    for (let i = 0; i < this._shadowLights.length; ++i) {
      let light = this._shadowLights[i];
      let view = _a16_shadow_lightViewProjs[i];
      if (!view) {
        view = _a16_shadow_lightViewProjs[i] = new Float32Array(_a64_shadow_lightViewProj.buffer, i * 64, 16);
      }
      Mat4.toArray(view, light.viewProjMatrix);

      let index = i*4;
      shadowInfo[index] = light.shadowMinDepth;
      shadowInfo[index+1] = light.shadowMaxDepth;
      shadowInfo[index+2] = light._shadowResolution;
      shadowInfo[index+3] = light.shadowDarkness;
    }

    this._device.setUniform(`cc_shadow_lightViewProjMatrix`, _a64_shadow_lightViewProj);
    this._device.setUniform(`cc_shadow_info`, shadowInfo);
    // this._device.setUniform(`cc_frustumEdgeFalloff_${index}`, light.frustumEdgeFalloff);
  }

  _sortItems (items) {
    // sort items
    items.sort((a, b) => {
      // if (a.layer !== b.layer) {
      //   return a.layer - b.layer;
      // }

      if (a.passes.length !== b.passes.length) {
        return a.passes.length - b.passes.length;
      }

      return a.sortKey - b.sortKey;
    });
  }

  _shadowStage (view, items) {
    // update rendering
    this._submitShadowStageUniforms(view);

    // this._sortItems(items);

    // draw it
    for (let i = 0; i < items.length; ++i) {
      let item = items.data[i];
      if (item.effect.getDefine('CC_CASTING_SHADOW')) {
        this._draw(item);
      }
    }
  }

  _drawItems (view, items) {
    let shadowLights = this._shadowLights;
    if (shadowLights.length === 0 && this._numLights === 0) {
      for (let i = 0; i < items.length; ++i) {
        let item = items.data[i];
        this._draw(item);
      }
    }
    else {
      for (let i = 0; i < items.length; ++i) {
        let item = items.data[i];

        for (let shadowIdx = 0; shadowIdx < shadowLights.length; ++shadowIdx) {
          this._device.setTexture('cc_shadow_map_'+shadowIdx, shadowLights[shadowIdx].shadowMap, this._allocTextureUnit());
        }

        this._draw(item);
      }
    }
  }

  _opaqueStage (view, items) {
    view.getPosition(_camPos);

    // update uniforms
    this._device.setUniform('cc_matView', Mat4.toArray(_a16_view, view._matView));
    this._device.setUniform('cc_matViewInv', Mat4.toArray(_a16_view_inv, view._matViewInv));
    this._device.setUniform('cc_matProj', Mat4.toArray(_a16_proj, view._matProj));
    this._device.setUniform('cc_matViewProj', Mat4.toArray(_a16_viewProj, view._matViewProj));
    this._device.setUniform('cc_cameraPos', Vec4.toArray(_a4_camPos, _camPos));

    // update rendering
    this._submitLightsUniforms();
    this._submitOtherStagesUniforms();

    this._drawItems(view, items);
  }

  _transparentStage (view, items) {
    view.getPosition(_camPos);
    view.getForward(_camFwd);

    // update uniforms
    this._device.setUniform('cc_matView', Mat4.toArray(_a16_view, view._matView));
    this._device.setUniform('cc_matViewInv', Mat4.toArray(_a16_view_inv, view._matViewInv));
    this._device.setUniform('cc_matProj', Mat4.toArray(_a16_proj, view._matProj));
    this._device.setUniform('cc_matViewProj', Mat4.toArray(_a16_viewProj, view._matViewProj));
    this._device.setUniform('cc_cameraPos', Vec4.toArray(_a4_camPos, _camPos));

    this._submitLightsUniforms();
    this._submitOtherStagesUniforms();

    // calculate zdist
    for (let i = 0; i < items.length; ++i) {
      let item = items.data[i];

      // TODO: we should use mesh center instead!
      item.node.getWorldPosition(_v3_tmp1);

      Vec3.sub(_v3_tmp1, _v3_tmp1, _camPos);
      item.sortKey = -Vec3.dot(_v3_tmp1, _camFwd);
    }

    this._sortItems(items);
    this._drawItems(view, items);
  }
}
